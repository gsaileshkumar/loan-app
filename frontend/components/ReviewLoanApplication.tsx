import { FormEvent } from "react";
import { FormData } from "../pages";

interface ReviewProps {
  formData?: FormData;
  onSubmit: (event: FormEvent) => void;
}

const ReviewLoanApplication: React.FunctionComponent<ReviewProps> = ({
  formData,
  onSubmit,
}) => {
  if (!formData) {
    return (
      <span className="appearance-none block w-full  text-gray-700 border rounded py-3  mb-3 leading-tight focus:outline-none focus:bg-white">
        No form data to review
      </span>
    );
  }

  const { email, name, year, loanAmount, accountingProvider } = formData;
  return (
    <form
      className="w-full max-w-lg bg-gray-100 p-5 rounded h-full"
      onSubmit={onSubmit}
    >
      <h2 className="w-full px-3 font-bold py-3 text-3xl">
        Review Loan Application
      </h2>
      <div className="w-full px-3 mb-6 md:mb-0">
        <label
          className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
          htmlFor="email"
        >
          Email
        </label>
        <span
          className="appearance-none block w-full  text-gray-700  rounded py-3  mb-3 leading-tight focus:outline-none focus:bg-white"
          id="email"
        >
          {email}
        </span>
      </div>
      <div className="w-full px-3 mb-6 md:mb-0">
        <label
          className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
          htmlFor="name"
        >
          Business name
        </label>
        <span
          className="appearance-none block w-full  text-gray-700  rounded py-3  mb-3 leading-tight focus:outline-none focus:bg-white"
          id="name"
        >
          {name}
        </span>
      </div>
      <div className="w-full px-3 mb-6 md:mb-0">
        <label
          className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
          htmlFor="email"
        >
          Year established
        </label>
        <span
          className="appearance-none block w-full  text-gray-700  rounded py-3  mb-3 leading-tight focus:outline-none focus:bg-white"
          id="year"
        >
          {year}
        </span>
      </div>
      <div className="w-full px-3 mb-6 md:mb-0">
        <label
          className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
          htmlFor="loanAmount"
        >
          Loan Amount
        </label>
        <span
          className="appearance-none block w-full  text-gray-700 rounded py-3  mb-3 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
          id="loanAmount"
        >
          {loanAmount}
        </span>
      </div>
      <div className="w-full px-3 mb-6 md:mb-0">
        <label
          className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
          htmlFor="accountingProvider"
        >
          Accounting provider
        </label>
        <div className="relative">
          <span
            className="appearance-none block w-full  text-gray-700  rounded py-3  mb-3 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
            id="accountingProvider"
          >
            {accountingProvider}
          </span>
        </div>
      </div>
      <div className="w-full px-3 mb-6 md:mb-0 flex justify-evenly">
        <button
          className="bg-gray-300 hover:bg-gray-400 text-gray-700 font-bold mt-4 py-2 px-4 rounded inline-flex items-center"
          type="submit"
        >
          <span>Submit</span>
        </button>
      </div>
    </form>
  );
};

export default ReviewLoanApplication;
