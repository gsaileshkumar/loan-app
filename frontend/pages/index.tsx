import fileDownload from "js-file-download";
import type { NextPage } from "next";
import Head from "next/head";
import { FormEvent, useState } from "react";
import LoanApplicationForm from "../components/LoanApplicationForm";
import ReviewLoanApplication from "../components/ReviewLoanApplication";
import SuccessPage from "../components/SuccessPage";
import api from "../utils/api";

type AppStateType = "EDIT" | "REVIEW" | "SUBMITTED";
export type AccountingProviderType = "xero" | "myob";
export type FormData = {
  email: string;
  name: string;
  year: number;
  loanAmount: number;
  accountingProvider: AccountingProviderType;
};
type ApprovalData = {
  approved: boolean;
  amount: number;
};

const LoanApp: NextPage = () => {
  const [state, setState] = useState<AppStateType>("EDIT");
  const [data, setData] = useState<FormData>();
  const [approvalData, setApprovalData] = useState<ApprovalData>();

  const getBalanceSheet = async (formData: FormData) => {
    const { accountingProvider, email } = formData;
    return await api()
      .get(`balanceSheet`, {
        accountingProvider,
        email,
      })
      .then((res) => fileDownload(res.data, "balance_sheet.csv"));
  };

  const onReviewHandler = (formData: FormData) => {
    setData(formData);
    setState("REVIEW");
  };

  const onSubmitHandler = async (event: FormEvent) => {
    event.preventDefault();
    return await api()
      .post(`submit`, data)
      .then((res) => res.data.data)
      .then((data) => {
        setState("SUBMITTED");
        setApprovalData(data);
      });
  };

  const onReset = () => {
    setState("EDIT");
  };

  return (
    <div className="bg-gray-200 h-screen">
      <Head>
        <title>Loan App</title>
        <meta name="description" content="Loan application" />
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <main className="flex justify-center items-center h-full">
        {state === "EDIT" ? (
          <LoanApplicationForm
            getBalanceSheet={getBalanceSheet}
            onReview={onReviewHandler}
          />
        ) : null}
        {state === "REVIEW" ? (
          <ReviewLoanApplication formData={data} onSubmit={onSubmitHandler} />
        ) : null}
        {state === "SUBMITTED" ? (
          <SuccessPage
            approved={approvalData?.approved}
            amount={approvalData?.amount}
            onReset={onReset}
          />
        ) : null}
      </main>
    </div>
  );
};

export default LoanApp;
