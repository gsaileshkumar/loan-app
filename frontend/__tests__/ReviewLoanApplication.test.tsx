import React from "react";
import renderer from "react-test-renderer";
import ReviewLoanApplication from "../components/ReviewLoanApplication";
import { FormData } from "../pages";

const TEST_DATA: FormData = {
  email: "abc@cdf.com",
  name: "qwerty",
  year: 2021,
  loanAmount: 20000,
  accountingProvider: "myob",
};

it("renders correctly", () => {
  const onSubmitHandler = jest.fn();
  const tree = renderer
    .create(
      <ReviewLoanApplication formData={TEST_DATA} onSubmit={onSubmitHandler} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
