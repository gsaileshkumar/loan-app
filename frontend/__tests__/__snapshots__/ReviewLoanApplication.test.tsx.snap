// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  className="w-full max-w-lg bg-gray-100 p-5 rounded h-full"
  onSubmit={[MockFunction]}
>
  <h2
    className="w-full px-3 font-bold py-3 text-3xl"
  >
    Review Loan Application
  </h2>
  <div
    className="w-full px-3 mb-6 md:mb-0"
  >
    <label
      className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
      htmlFor="email"
    >
      Email
    </label>
    <span
      className="appearance-none block w-full text-gray-700 rounded py-3 mb-3 leading-tight focus:outline-none focus:bg-white"
      id="email"
    >
      abc@cdf.com
    </span>
  </div>
  <div
    className="w-full px-3 mb-6 md:mb-0"
  >
    <label
      className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
      htmlFor="name"
    >
      Business name
    </label>
    <span
      className="appearance-none block w-full text-gray-700 rounded py-3 mb-3 leading-tight focus:outline-none focus:bg-white"
      id="name"
    >
      qwerty
    </span>
  </div>
  <div
    className="w-full px-3 mb-6 md:mb-0"
  >
    <label
      className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
      htmlFor="email"
    >
      Year established
    </label>
    <span
      className="appearance-none block w-full text-gray-700 rounded py-3 mb-3 leading-tight focus:outline-none focus:bg-white"
      id="year"
    >
      2021
    </span>
  </div>
  <div
    className="w-full px-3 mb-6 md:mb-0"
  >
    <label
      className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
      htmlFor="loanAmount"
    >
      Loan Amount
    </label>
    <span
      className="appearance-none block w-full text-gray-700 rounded py-3 mb-3 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
      id="loanAmount"
    >
      20000
    </span>
  </div>
  <div
    className="w-full px-3 mb-6 md:mb-0"
  >
    <label
      className="block uppercase tracking-wide text-gray-700 text-xs font-bold mb-2"
      htmlFor="accountingProvider"
    >
      Accounting provider
    </label>
    <div
      className="relative"
    >
      <span
        className="appearance-none block w-full text-gray-700 rounded py-3 mb-3 leading-tight focus:outline-none focus:bg-white focus:border-gray-500"
        id="accountingProvider"
      >
        myob
      </span>
    </div>
  </div>
  <div
    className="w-full px-3 mb-6 md:mb-0 flex justify-evenly"
  >
    <button
      className="bg-gray-300 hover:bg-gray-400 text-gray-700 font-bold mt-4 py-2 px-4 rounded inline-flex items-center"
      type="submit"
    >
      <span>
        Submit
      </span>
    </button>
  </div>
</form>
`;
